File: python/ctxshrink/reducers/prose.py
"""Prose reducer: trims verbose phrasing while keeping the claim it makes.

Deterministic phrase substitution, not paraphrase. Every rule replaces a
padded construction with its plain-English equivalent (filler removal, passive
throat-clearing, wordy connectives) so the transform is auditable line by line
and never invents or drops a claim. Meaning-bearing content is never removed;
only the words around it are trimmed.
"""

from __future__ import annotations

import re
from typing import Optional

from ..levels import LEVEL_AGGRESSIVE, LEVEL_CONSERVATIVE
from ..safety import SafetyPolicy, check_safety

__all__ = ["reduce", "PHRASE_RULES", "FILLER_SENTENCES"]

# (pattern, replacement) applied at LEVEL_CONSERVATIVE and above.
# Ordered longest-match-first so overlapping phrases resolve predictably.
PHRASE_RULES: tuple = (
    (r"\bin order to\b", "to"),
    (r"\bdue to the fact that\b", "because"),
    (r"\bowing to the fact that\b", "because"),
    (r"\bat this (?:point|moment) in time\b", "now"),
    (r"\bfor the purpose of\b", "to"),
    (r"\bin the event that\b", "if"),
    (r"\bin the case that\b", "if"),
    (r"\bwith regard to\b", "about"),
    (r"\bwith respect to\b", "about"),
    (r"\bin relation to\b", "about"),
    (r"\bis able to\b", "can"),
    (r"\bare able to\b", "can"),
    (r"\bwas able to\b", "could"),
    (r"\bwere able to\b", "could"),
    (r"\bthe majority of\b", "most"),
    (r"\ba (?:large|significant) number of\b", "many"),
    (r"\ba number of\b", "several"),
    (r"\beach and every\b", "each"),
    (r"\bwhether or not\b", "whether"),
    (r"\bin spite of the fact that\b", "although"),
    (r"\bdespite the fact that\b", "although"),
    (r"\bit is (?:important|worth noting) to (?:note|mention) that\b", ""),
    (r"\bit should be noted that\b", ""),
    (r"\bplease note that\b", "note:"),
    (r"\bplease be advised that\b", ""),
    (r"\bas a matter of fact\b", ""),
    (r"\bfor all intents and purposes\b", ""),
    (r"\bat the end of the day\b", ""),
    (r"\bneeds to be able to\b", "must"),
    (r"\bhas the ability to\b", "can"),
    (r"\bmake use of\b", "use"),
    (r"\butilize\b", "use"),
    (r"\butilization\b", "use"),
    (r"\bcommence\b", "start"),
    (r"\bterminate\b", "end"),
    (r"\bsubsequently\b", "then"),
    (r"\bprior to\b", "before"),
    (r"\bin the near future\b", "soon"),
    (r"\bon a regular basis\b", "regularly"),
    (r"\bin a timely manner\b", "promptly"),
    (r"\bvery unique\b", "unique"),
)

# Whole leading sentences/clauses that add no information; dropped if they
# start a sentence, kept if they are the entire content (avoid emptying text).
FILLER_SENTENCES: tuple = (
    r"^as (?:you|we) (?:can see|know),?\s*",
    r"^it goes without saying that\s*",
    r"^needless to say,?\s*",
    r"^generally speaking,?\s*",
    r"^basically,?\s*",
    r"^essentially,?\s*",
    r"^simply put,?\s*",
    r"^in summary,?\s*",
    r"^to summarize,?\s*",
    r"^in conclusion,?\s*",
)

_PHRASE_COMPILED = [(re.compile(p, re.IGNORECASE), r) for p, r in PHRASE_RULES]
_FILLER_COMPILED = [re.compile(p, re.IGNORECASE) for p in FILLER_SENTENCES]
_WHITESPACE_RUN_RE = re.compile(r"[ \t]{2,}")
_BLANK_RUN_RE = re.compile(r"\n{3,}")


def _apply_phrase_rules(text: str) -> str:
    out = text
    for pattern, replacement in _PHRASE_COMPILED:
        out = pattern.sub(replacement, out)
    return out


def _apply_filler_removal(text: str) -> str:
    lines = text.split("\n")
    result = []
    for line in lines:
        trimmed = line
        for pattern in _FILLER_COMPILED:
            candidate = pattern.sub("", trimmed)
            if candidate.strip():
                trimmed = candidate
        result.append(trimmed)
    return "\n".join(result)


def _tidy_whitespace(text: str) -> str:
    text = _WHITESPACE_RUN_RE.sub(" ", text)
    text = _BLANK_RUN_RE.sub("\n\n", text)
    lines = [line.rstrip() for line in text.split("\n")]
    return "\n".join(lines)


def reduce(
    text: str,
    level: int = LEVEL_CONSERVATIVE,
    policy: Optional[SafetyPolicy] = None,
) -> Optional[str]:
    if level < LEVEL_CONSERVATIVE or not text.strip():
        return None

    out = _apply_phrase_rules(text)
    if level >= LEVEL_AGGRESSIVE:
        out = _apply_filler_removal(out)
    out = _tidy_whitespace(out).strip("\n")

    if out == text or not out:
        return None

    policy = policy or SafetyPolicy(max_drop_ratio=0.5 if level < LEVEL_AGGRESSIVE else 0.7)
    if check_safety(text, out, policy):
        return None
    return out

File: js/src/reducers/prose.js
/**
 * Prose reducer: trims verbose phrasing while keeping the claim it makes.
 *
 * Deterministic phrase substitution, not paraphrase. Every rule replaces a
 * padded construction with its plain-English equivalent (filler removal,
 * passive throat-clearing, wordy connectives) so the transform is auditable
 * line by line and never invents or drops a claim. Meaning-bearing content
 * is never removed; only the words around it are trimmed.
 */

import { LEVEL_AGGRESSIVE, LEVEL_CONSERVATIVE } from "../levels.js";
import { SafetyPolicy, checkSafety } from "../safety.js";

// [pattern, replacement] applied at LEVEL_CONSERVATIVE and above. Ordered
// longest-match-first so overlapping phrases resolve predictably.
const PHRASE_RULES = [
  [/\bin order to\b/gi, "to"],
  [/\bdue to the fact that\b/gi, "because"],
  [/\bowing to the fact that\b/gi, "because"],
  [/\bat this (?:point|moment) in time\b/gi, "now"],
  [/\bfor the purpose of\b/gi, "to"],
  [/\bin the event that\b/gi, "if"],
  [/\bin the case that\b/gi, "if"],
  [/\bwith regard to\b/gi, "about"],
  [/\bwith respect to\b/gi, "about"],
  [/\bin relation to\b/gi, "about"],
  [/\bis able to\b/gi, "can"],
  [/\bare able to\b/gi, "can"],
  [/\bwas able to\b/gi, "could"],
  [/\bwere able to\b/gi, "could"],
  [/\bthe majority of\b/gi, "most"],
  [/\ba (?:large|significant) number of\b/gi, "many"],
  [/\ba number of\b/gi, "several"],
  [/\beach and every\b/gi, "each"],
  [/\bwhether or not\b/gi, "whether"],
  [/\bin spite of the fact that\b/gi, "although"],
  [/\bdespite the fact that\b/gi, "although"],
  [/\bit is (?:important|worth noting) to (?:note|mention) that\b/gi, ""],
  [/\bit should be noted that\b/gi, ""],
  [/\bplease note that\b/gi, "note:"],
  [/\bplease be advised that\b/gi, ""],
  [/\bas a matter of fact\b/gi, ""],
  [/\bfor all intents and purposes\b/gi, ""],
  [/\bat the end of the day\b/gi, ""],
  [/\bneeds to be able to\b/gi, "must"],
  [/\bhas the ability to\b/gi, "can"],
  [/\bmake use of\b/gi, "use"],
  [/\butilize\b/gi, "use"],
  [/\butilization\b/gi, "use"],
  [/\bcommence\b/gi, "start"],
  [/\bterminate\b/gi, "end"],
  [/\bsubsequently\b/gi, "then"],
  [/\bprior to\b/gi, "before"],
  [/\bin the near future\b/gi, "soon"],
  [/\bon a regular basis\b/gi, "regularly"],
  [/\bin a timely manner\b/gi, "promptly"],
  [/\bvery unique\b/gi, "unique"],
];

// Whole leading sentences/clauses that add no information; dropped if they
// start a sentence, kept if they are the entire content (avoid emptying
// text).
const FILLER_SENTENCES = [
  /^as (?:you|we) (?:can see|know),?\s*/i,
  /^it goes without saying that\s*/i,
  /^needless to say,?\s*/i,
  /^generally speaking,?\s*/i,
  /^basically,?\s*/i,
  /^essentially,?\s*/i,
  /^simply put,?\s*/i,
  /^in summary,?\s*/i,
  /^to summarize,?\s*/i,
  /^in conclusion,?\s*/i,
];

const WHITESPACE_RUN_RE = /[ \t]{2,}/g;
const BLANK_RUN_RE = /\n{3,}/g;

function applyPhraseRules(text) {
  let out = text;
  for (const [pattern, replacement] of PHRASE_RULES) {
    out = out.replace(pattern, replacement);
  }
  return out;
}

function applyFillerRemoval(text) {
  return text
    .split("\n")
    .map((line) => {
      let trimmed = line;
      for (const pattern of FILLER_SENTENCES) {
        const candidate = trimmed.replace(pattern, "");
        if (candidate.trim()) trimmed = candidate;
      }
      return trimmed;
    })
    .join("\n");
}

function tidyWhitespace(text) {
  text = text.replace(WHITESPACE_RUN_RE, " ");
  text = text.replace(BLANK_RUN_RE, "\n\n");
  return text
    .split("\n")
    .map((line) => line.replace(/\s+$/, ""))
    .join("\n");
}

function reduce(text, level = LEVEL_CONSERVATIVE, policy = null) {
  if (level < LEVEL_CONSERVATIVE || !text.trim()) return null;

  let out = applyPhraseRules(text);
  if (level >= LEVEL_AGGRESSIVE) out = applyFillerRemoval(out);
  out = tidyWhitespace(out).replace(/\n+$/, "");

  if (out === text || !out) return null;

  const effective =
    policy ?? new SafetyPolicy({ maxDropRatio: level < LEVEL_AGGRESSIVE ? 0.5 : 0.7 });
  if (checkSafety(text, out, effective).length > 0) return null;
  return out;
}

export { reduce, PHRASE_RULES, FILLER_SENTENCES };